### PR TITLE
[linux] add Pipewire audio output support

### PR DESCRIPTION
**Pull request by lrusak**
This PR adds initial support for a native [pipewire](https://pipewire.org/) audio sink.

I went for the full fat c++ wrapper style. This may seem overkill but it adds flexibility for the future. It also makes things nicely organised and (IMO) easier to read.

My idea is to have central access to the implemented pipewire interface. Pipewire like pulseaudio relies on a threaded mainloop and callbacks. These callback tell us when pipewire is finished initialising, enumerating, etc. Passing the c++ classes as the userdata for these static callbacks works nicely.

I've set the required version to 0.3.24 as that's the current version on my workstation. This likely works on earlier versions but Pipewire development is very active and I think it's fine to require a current version.

Fedora plans to ship Pipewire as a pulseaudio replacement for the upcoming F34 version. Though pipewire implements a pulseaudio server to provide support for existing application it would be good to have a native implementation.

I've currently put pipewire registration after pulseaudio because the pulseaudio sink is more feature rich (passthrough, volume control). So currently you either need to build without pulseaudio or run using `KODI_AE_SINK=PIPEWIRE ./kodi.bin`

Currently supported:
 - selecting specific audio output devices
 - selecting the default output device 
 - enumerating available outputs when devices are added/removed
 - enumerating output device channel mapping

Not supported:
 - passthrough support (pipewire limitation currently)
 - volume control (to be added in the future) (volume control works fine through kodi sw volume)

Things that could possibly use improvement:
 - I don't like the (effective) global in the AESink. I did this to avoid having to _restart_ pipewire 3 times (registration, enumeration, initialization). I don't really see a way around this though and the global is scoped to the namespace.
 - I'm not 100% sure about the main thread signaling vs the pipewire core sync methods. Something to think about but it works for now
 - enumerating the sink properties is a bit ugly, we only need the channel map for now but I also enumerate the hw formats and rates (but don't use them as pipewire uses float anyways).

I've separated the commits for easier review but can squash them if needed for (possible) merging.

Please comment on anything and I'll try to explain why it was done like that.

PS. It always amazes me how kodi has the ability to add a new feature an have a huge amount of additional lines and practically zero deletions :1st_place_medal:

### CODE REVIEW

**Review by fritsch — COMMENTED** [review]
I don't think that will work in a robust way as it is right now.

File: xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+
+  void* buffer = data[0] + offset * m_format.m_frameSize;
+
+  std::memcpy(spaData->data, buffer, length);
+
+  spaData->chunk->offset = 0;
+  spaData->chunk->stride = m_format.m_frameSize;
+  spaData->chunk->size = length;
+
+  stream->QueueBuffer(pwBuffer);
```
[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp:436]: blocking is done in the deque

File: xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    }
+
+    id = target->first;
+  }
+
+  stream = std::make_shared<PIPEWIRE::CPipewireStream>(core->Get());
+
+  stream->AddListener(pipewire.get());
+
+  m_latency = 20; // ms
```
[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp:311]: I actually started with 100ms of latency but there was a noticeable delay in the gui sounds. I'm not sure if there was a combination of miscalculations in `GetDelay and `GetCacheTotal` that caused that or what. So i tried lowering it to 20ms and it seems to work just fine. I don't get any warnings of underrun from kodi or pipewire (that I noticed). It's probably something to play around with more though as my understanding of how it all works is still limited.

[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp:311]:
```
2021-04-07 13:19:38.084 T:2467572   DEBUG <general>: CActiveAESink::OpenSink - trying to open device PIPEWIRE:Default
2021-04-07 13:19:38.084 T:2467572   DEBUG <general>: CAESinkPipewire::Initialize - rate: 96000
2021-04-07 13:19:38.084 T:2467572   DEBUG <general>: CAESinkPipewire::Initialize - channels: 2
2021-04-07 13:19:38.084 T:2467572   DEBUG <general>: CAESinkPipewire::Initialize - format: f32
2021-04-07 13:19:38.084 T:2467572   DEBUG <general>: CAESinkPipewire::Initialize - samplesize: 4
2021-04-07 13:19:38.084 T:2467572   DEBUG <general>: CAESinkPipewire::Initialize - framesize: 8
2021-04-07 13:19:38.084 T:2467572   DEBUG <general>: CAESinkPipewire::Initialize - latency: 1920/96000 (0.020s)
2021-04-07 13:19:38.084 T:2467572   DEBUG <general>: CPipewireStream::StateChanged - stream state changed unconnected -> connecting
2021-04-07 13:19:38.085 T:2467572   DEBUG <general>: CAESinkPipewire::Initialize - waiting
2021-04-07 13:19:38.085 T:2467560   DEBUG <general>: CPipewireNode::Info - node 39 changed
2021-04-07 13:19:38.085 T:2467560   DEBUG <general>: CPipewireStream::StateChanged - stream state changed connecting -> paused
2021-04-07 13:19:38.086 T:2467572   DEBUG <general>: CAESinkPipewire::Initialize - initialized
2021-04-07 13:19:38.086 T:2467572   DEBUG <general>: CActiveAESink::OpenSink - PIPEWIRE Initialized:
2021-04-07 13:19:38.086 T:2467572   DEBUG <general>:   Output Device : Default
2021-04-07 13:19:38.086 T:2467572   DEBUG <general>:   Sample Rate   : 96000
2021-04-07 13:19:38.086 T:2467572   DEBUG <general>:   Sample Format : AE_FMT_FLOAT
2021-04-07 13:19:38.086 T:2467572   DEBUG <general>:   Channel Count : 2
2021-04-07 13:19:38.086 T:2467572   DEBUG <general>:   Channel Layout: FL, FR
2021-04-07 13:19:38.086 T:2467572   DEBUG <general>:   Frames        : 1920
2021-04-07 13:19:38.086 T:2467572   DEBUG <general>:   Frame Size    : 8
```

[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp:311]:
```
Apr 07 13:19:41 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999983 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (0.000004 0.000007 0.000010)
Apr 07 13:19:44 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999976 bw:0.128000 thr:512 del:513 target:512 err:1.000000 (0.000023 0.000016 0.000008)
Apr 07 13:19:47 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999986 bw:0.128000 thr:512 del:513 target:512 err:1.000000 (0.000011 0.000009 0.000005)
Apr 07 13:19:50 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999996 bw:0.128000 thr:512 del:510 target:512 err:-2.000000 (-0.000021 -0.000005 0.000009)
Apr 07 13:19:53 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999995 bw:0.128000 thr:512 del:513 target:512 err:1.000000 (0.000004 -0.000001 0.000006)
Apr 07 13:19:56 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:1.000003 bw:0.128000 thr:512 del:511 target:512 err:-1.000000 (-0.000007 -0.000008 0.000004)
Apr 07 13:19:59 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999992 bw:0.128000 thr:512 del:513 target:512 err:1.000000 (0.000000 0.000000 0.000008)
Apr 07 13:20:02 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999991 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (0.000007 0.000006 0.000003)
Apr 07 13:20:05 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999991 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (0.000001 0.000004 0.000005)
Apr 07 13:20:08 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999968 bw:0.128000 thr:512 del:510 target:512 err:-2.000000 (0.000022 0.000026 0.000006)
Apr 07 13:20:11 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999992 bw:0.128000 thr:512 del:513 target:512 err:1.000000 (0.000008 0.000000 0.000008)
Apr 07 13:20:14 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999999 bw:0.128000 thr:512 del:510 target:512 err:-2.000000 (-0.000007 -0.000005 0.000006)
Apr 07 13:20:17 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999979 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (0.000010 0.000012 0.000009)
Apr 07 13:20:20 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999988 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (0.000013 0.000010 0.000002)
Apr 07 13:20:23 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999995 bw:0.128000 thr:512 del:510 target:512 err:-2.000000 (-0.000011 -0.000003 0.000008)
Apr 07 13:20:26 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:1.000003 bw:0.128000 thr:512 del:511 target:512 err:-1.000000 (-0.000003 -0.000004 0.000001)
Apr 07 13:20:29 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999993 bw:0.128000 thr:512 del:513 target:512 err:1.000000 (0.000004 -0.000001 0.000008)
Apr 07 13:20:32 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999985 bw:0.128000 thr:512 del:510 target:512 err:-2.000000 (0.000001 0.000007 0.000009)
Apr 07 13:20:35 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:1.000002 bw:0.128000 thr:512 del:511 target:512 err:-1.000000 (-0.000005 -0.000006 0.000004)
Apr 07 13:20:38 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999978 bw:0.128000 thr:512 del:513 target:512 err:1.000000 (0.000019 0.000013 0.000009)
Apr 07 13:20:41 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:1.000011 bw:0.128000 thr:512 del:513 target:512 err:1.000000 (-0.000010 -0.000017 0.000005)
Apr 07 13:20:44 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999998 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (0.000004 -0.000003 0.000005)
Apr 07 13:20:47 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999990 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (0.000005 0.000005 0.000005)
Apr 07 13:20:50 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999996 bw:0.128000 thr:512 del:510 target:512 err:-2.000000 (-0.000014 -0.000004 0.000008)
Apr 07 13:20:53 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999986 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (0.000012 0.000007 0.000007)
Apr 07 13:20:56 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999992 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (-0.000006 -0.000008 0.000016)
Apr 07 13:20:59 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999997 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (-0.000002 -0.000004 0.000007)
Apr 07 13:21:02 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999996 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (-0.000000 -0.000000 0.000004)
Apr 07 13:21:05 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999994 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (-0.000001 -0.000000 0.000006)
Apr 07 13:21:08 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999995 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (-0.000002 -0.000002 0.000007)
Apr 07 13:21:11 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999994 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (-0.000001 -0.000001 0.000007)
Apr 07 13:21:14 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999993 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (0.000000 0.000000 0.000007)
Apr 07 13:21:17 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999994 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (-0.000000 -0.000001 0.000006)
Apr 07 13:21:20 nuc8i7hnk pipewire[2265010]: alsa-pcm 0x55b1154091c8: follower:0 match:0 rate:0.999995 bw:0.128000 thr:512 del:512 target:512 err:0.000000 (-0.000003 -0.000001 0.000006)
```

[the remark this answers is not in the document]
Comment (fritsch) [reply xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp:311]: Basically the 20 ms you set here, you also use as frames. Means AE sends you 20 ms frames (mostly) and while doing so, the Buffer runs dry.

Try: periods = 50 ms and buffer (in your words delay=200ms) or something. The TotalCache is then (after initially being 50,100,150,200 most of the time besides you run dry).

Currently it works by luck I guess. When your sink does not come back withing 20ms you run dry.

**Comment by lrusak** [comment]
adding @wsnipex for cmake review

File: cmake/modules/FindPipewire.cmake [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+# This will define the following variables::
+#
+#  PIPEWIRE_FOUND - system has the Pipewire library
+#  PIPEWIRE_INCLUDE_DIRS - the Pipewire include directory
+#  PIPEWIRE_LIBRARIES - the libraries needed to use Pipewire
+#  PIPEWIRE_DEFINITIONS - the definitions needed to use Pipewire
+#
+
+if(PKG_CONFIG_FOUND)
+  pkg_check_modules(PC_PIPEWIRE libpipewire-0.3>=0.3.24 QUIET)
```
Comment (wsnipex) [inline cmake/modules/FindPipewire.cmake:15]: you should probably check for libspa-0.2 here as well

Comment (lrusak) [reply cmake/modules/FindPipewire.cmake:15]: Added, thanks!

**Review by wsnipex — APPROVED** [review]
ack for cmake stuff

**Comment by lrusak** [comment]
I'll probably just end up merging this soon than later. Any other changes and fixes can be addressed in follow up PR's

**Comment by asavah** [comment]
Finally had some time to give this PR a test run.
It works for 2 minutes ... then it fails. Log below.

Relevant setup details:
pipewire: 0.3.27
bluetooth headset (OnePlus Bullets), works like a charm with Kodi and pipewire-pulse, No issues at all.

Debug log https://paste.kodi.tv/ababusikek.kodi

What has been done:
Start Kodi with headset already connected.
UI sounds work.
Try playing a video file - all ok, audio worked.
Quickly scroll down a folder with a few dozens of files.
Ui sounds stopped reproducing.
Tried playing a video file - started playing and almost inmediately stalled.
Stopped playback.
Tried quitting Kodi - Kodi hangs, becomes totally irresponsive.
Killed Kodi with SIGKILL.

The only thing I see in syslog that might be relevant:
```
May 09 22:22:30 asusb450eg pipewire[588]: (bluez_output.98_09_CF_F1_8B_E4.a2dp-sink-38) client missed 1 wakeups
```

**Comment by lrusak** [comment]
There is no special handling for Bluetooth devices like the pulse audio sink has. I don't really intend on changing that for this PR either. If pulse works for you please just use that.

Thanks for testing otherwise.

**Comment by asavah** [comment]
The whole point of pipewire is that pipewire should take care of whatever audio stuff a user might have.
There should be no need for any special treatment for bluetooth or whatever on the client side.
Given that with this PR Kodi halts to the point that it need to be killed IMO it clearly shows that something might be very wrong with the code on Kodi side.
I mentioned that pipewire-pulse works OK only to point out that this setup has been working fine and there are no issues with it.
The whole point of testing PRs is to provide some feedback before potentially broken code hits the repo.

**Comment by fritsch** [comment]
The buffer and periods handling as well as the delay is still not correct.

**Comment by L2501** [comment]
missing pipewire/keys.h in this scope?

https://github.com/lrusak/xbmc/blob/5cda6a0b423564827119df14f2902985b2c767e3/xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp#L315-L320

**Pull request merged**